### PR TITLE
DISPATCH-209 : add a linear 3-router linkroute test

### PR DESCRIPTION
**Pull request by mgoulish**
test #4 -- 3-router link-route

### CODE REVIEW

**Comment by mgoulish** [comment]
I had an 8-in-100 failure rate suddenly appear on test #3 -- so... I made on_sendable handling smarter and that gave me 4-in-100 failure rate -- all on teardown only.    I then changed a little bit how all four tests shut down -- borrowing an interesting line from another test (null out the timer after clearing it) -- and now I have  0 failures in 400 iterations -- also no teardown failures.

**Comment by mgoulish** [comment]
WAIT

I just had a failure on my tests while running them in the context of a full "make check" -- investigating.....

File: tests/system_tests_three_routers.py [hunk 1]
```diff
@@ -96,10 +124,24 @@ def router ( name, connection_1, connection_2=None ):
                          'port': inter_router_port_A,
                          'verifyHostName': 'no'
                        }
+                   ),
+                   ( 'linkRoute',
+                     { 'prefix': cls.linkroute_prefix,
+                       'containerId': 'QDR.A',
```
Comment (ted-ross) [inline tests/system_tests_three_routers.py]: These containerId lines have no effect and are misleading.

File: tests/system_tests_three_routers.py [hunk 2]
```diff
@@ -335,36 +417,124 @@ def on_start ( self, event ):
...
+    def __init__ ( self, addr_listening_for_route_container, linkroute_prefix, qdstat_address, send_address ):
+        super(LinkRoute, self).__init__(prefetch=0)
+        self.addr_listening_for_route_container = addr_listening_for_route_container
+        self.send_address = send_address
+        self.linkroute_prefix = linkroute_prefix
+        self.qdstat_address = qdstat_address
+        self.error      = None
+        self.send_connection = None
+        self.recv_connection = None
+        self.false_statement = "This is not the message you are looking for."
```
Comment (ted-ross) [inline tests/system_tests_three_routers.py]: For anyone trying to understand how this test works, this text is misleading.

File: tests/system_tests_three_routers.py [hunk 3]
```diff
@@ -335,36 +417,124 @@ def on_start ( self, event ):
...
+
+
+    def on_start(self, event):
+        # Iff this timer expires, the test fails.
+        self.timer = event.reactor.schedule ( 5, Timeout(self) )
+
+        # Create the receiver only at start time.
+        # Sender-creation has to wait until we know that the address is available.
+        self.recv_connection = event.container.connect ( self.addr_listening_for_route_container )
+        self.receiver = event.container.create_receiver ( self.recv_connection, None, dynamic=True )
```
Comment (ted-ross) [inline tests/system_tests_three_routers.py]: Why is this receiver here?  It serves no purpose in the test.

File: tests/system_tests_three_routers.py [hunk 4]
```diff
@@ -335,36 +417,124 @@ def on_start ( self, event ):
...
+
+
+    def on_link_opening ( self, event ):
+        if event.receiver:
+            event.receiver.flow(1)
+
+
+    def on_connection_remote_open ( self, event ):
```
Comment (ted-ross) [inline tests/system_tests_three_routers.py]: Doing this blocking, synchronous operation in a reactive event handler is not a good practice.  The reactor is prevented from maintaining its connections for the entire duration of this function.
There is an example in system_tests_multi_tenancy (LinkRouteTest) that does what's needed to wait for the link-route to propagate within the reactive framework.

File: tests/system_tests_three_routers.py [hunk 5]
```diff
@@ -335,36 +417,124 @@ def on_start ( self, event ):
...
+        # Iff this timer expires, the test fails.
+        self.timer = event.reactor.schedule ( 5, Timeout(self) )
+
+        # Create the receiver only at start time.
+        # Sender-creation has to wait until we know that the address is available.
+        self.recv_connection = event.container.connect ( self.addr_listening_for_route_container )
+        self.receiver = event.container.create_receiver ( self.recv_connection, None, dynamic=True )
+
+
+    def on_link_opening ( self, event ):
```
Comment (ted-ross) [inline tests/system_tests_three_routers.py]: When handing an unsolicited incoming link-attach, this handler must copy the addresses from the remote to the local.  If you trace the operation of this test, you will see that the returning attach (from the initial attach to "myLinkRoute") does not contain an address.  This is a protocol violation and it is by luck that the link correctly transfers the message.
The system_tests_multi_tenancy/LinkRouteTest test can be used an an example of the correct way to handle a link-routed attach.
